### PR TITLE
Warn about ignored generic bounds in `for`

### LINKED ISSUE #42181: Bounds in HRTB are ignored
The syntax `where for<'b: 'a> ...` is permitted without warning, however the `: 'a` bound is not used to prove the identity; I would expect the following to compile:

```rust
trait Trait {}

impl Trait for &'static () {}

fn f<'a>(_: &'a ()) where for<'b: 'a> &'b () : Trait {}

fn main() {
    static X: &'static () = &();
    f::<'static>(X)
}
```

but it fails with

```
rustc 1.17.0 (56124baa9 2017-04-24)
error[E0277]: the trait bound `for<'b> &'b (): Trait` is not satisfied
 --> <anon>:9:5
  |
9 |     f::<'static>(X)
  |     ^^^^^^^^^^^^ the trait `for<'b> Trait` is not implemented for `&'b ()`
  |
  = help: the following implementations were found:
            <&'static () as Trait>
  = note: required by `f`

error: aborting due to previous error
```

### PR DESCRIPTION
**Pull request by RalfJung**
This adds a new lint to fix #42181. For consistency and to avoid code duplication, I also moved the existing "bounds in type aliases are ignored" here.

Questions to the reviewer:
* Is it okay to just remove a diagnostic error code like this? Should I instead keep the warning about type aliases where it is? The old code provided a detailed explanation of what's going on when asked, that information is now lost. On the other hand, `span_warn!` seems deprecated (after this patch, it has exactly one user left!).
* Did I miss any syntactic construct that can appear as `for` in the surface syntax? I covered function types (`for<'a> fn(...)`), generic traits (`for <'a> Fn(...)`, can appear both as bounds as as trait objects) and bounds (`for<'a> F: ...`).
* For the sake of backwards compatibility, this adds a warning, not an error. @nikomatsakis suggested an error in https://github.com/rust-lang/rust/issues/42181#issuecomment-306924389, but I feel that can only happen in a new epoch -- right?

Cc @eddyb

### CODE REVIEW

**Comment by petrochenkov** [comment]
>Is it okay to just remove a diagnostic error code like this?

It's okay. `src/librustc_typeck/diagnostics.rs` contains a list of commented out error codes at the very end of the file. `E0122` can be added to that list as well with a short note on why it was removed.

>On the other hand, span_warn! seems deprecated

Yes, hardcoded warnings should be avoided whenever possible (I don't think we should remove the mechanism itself though).

>Did I miss any syntactic construct that can appear as for in the surface syntax?

Everything looks covered.

>For the sake of backwards compatibility, this adds a warning, not an error. @nikomatsakis suggested an error in #42181 (comment), but I feel that can only happen in a new epoch -- right?

Crater can answer this! cc @aidanhs

File: src/test/ui/param-bounds-ignored.rs [hunk 1]
```diff
@@ -9,12 +9,14 @@
 // except according to those terms.
 
 // must-compile-successfully
+#![allow(dead_code, non_camel_case_types)]
```
Comment (petrochenkov) [inline src/test/ui/param-bounds-ignored.rs]: Could you `warn(ignored_generic_bounds)` and add `//~ WARN ...` annotations to the file?
It's very hard to see what is tested, what should be reported, and what shouldn't be reported based on `stderr` file alone.

File: src/test/ui/param-bounds-ignored.stderr [hunk 2]
```diff
@@ -1,18 +1,92 @@
...
+  --> $DIR/param-bounds-ignored.rs:19:1
+   |
+19 | type W2Vec<'b, T> where T: 'b = Vec<T>;
+   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+warning: bounds on generic parameters are ignored in higher-ranked function types (i.e., `for`)
+  --> $DIR/param-bounds-ignored.rs:39:8
+   |
+39 |     f: for<'xa, 'xb: 'xa> fn(&'xa i32, &'xb i32) -> &'xa i32)
+   |        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```
Comment (petrochenkov) [inline src/test/ui/param-bounds-ignored.stderr]: Spans could be more precise, AST has spans for generics and for the ignored bounds themselves.

**Comment by petrochenkov** [comment]
>Crater can answer this!

The warning should be forbid-by-default for cratering though.

**Comment by RalfJung** [comment]
> The warning should be forbid-by-default for cratering though.

So are you saying I should change this PR to make the warning err-by-default?

File: src/test/ui/param-bounds-ignored.stderr (hunk 2 above)
Comment (RalfJung) [reply src/test/ui/param-bounds-ignored.stderr]: I'm now changing this so the span points at the first ignored bound. Is there a way to "combine" spans? I found no span for all the bounds together, just the generic (which is the wrong place) and individual bounds.

File: src/test/ui/param-bounds-ignored.rs (hunk 1 above)
Comment (RalfJung) [reply src/test/ui/param-bounds-ignored.rs]: Ah, I didn't even know `//~ WARN` was a thing in ui tests. How do the two interact?

**Comment by RalfJung** [comment]
Turns out that even some of the compile-fail tests have these bounds in `for`. I'd be very surprised if crater wouldn't find tons of cases^^

File: src/test/ui/param-bounds-ignored.rs (hunk 1 above)
Comment (RalfJung) [reply src/test/ui/param-bounds-ignored.rs]: Done.

**Comment by RalfJung** [comment]
> E0122 can be added to that list as well with a short note on why it was removed.

Done.

**Comment by RalfJung** [comment]
There's a strange failure in `compile-fail/enum-discrim-too-small2.rs`: the compiler panics with the following backtrace:
```
thread 'rustc' panicked at 'Box<Any>', librustc_errors/lib.rs:535:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:206
   3: std::panicking::default_hook
             at libstd/panicking.rs:222
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:400
   5: std::panicking::begin_panic
             at ./src/libstd/panicking.rs:363
   6: rustc_errors::Handler::bug
             at librustc_errors/lib.rs:535
   7: <std::thread::local::LocalKey<T>>::with
             at librustc/session/mod.rs:1179
             at librustc/ty/context.rs:1600
             at librustc/ty/context.rs:1589
             at ./src/libstd/thread/local.rs:377
             at ./src/libstd/thread/local.rs:288
   8: rustc::ty::context::tls::with_opt
             at librustc/ty/context.rs:1585
             at librustc/ty/context.rs:1600
   9: rustc::session::opt_span_bug_fmt
             at librustc/session/mod.rs:1175
  10: rustc::session::bug_fmt
             at librustc/session/mod.rs:1159
  11: <rustc::ty::layout::LayoutCx<'tcx, rustc::ty::context::TyCtxt<'a, 'tcx, 'tcx>>>::layout_raw_uncached
             at librustc/ty/layout.rs:553
             at librustc/ty/layout.rs:1545
  12: rustc::ty::layout::layout_raw
             at librustc/ty/layout.rs:899
  13: rustc::ty::maps::<impl rustc::ty::maps::queries::layout_raw<'tcx>>::compute_result
             at librustc/ty/maps/plumbing.rs:396
  14: rustc::dep_graph::graph::DepGraph::with_task_impl
             at librustc/dep_graph/graph.rs:289
  15: rustc_errors::Handler::track_diagnostics
             at librustc/dep_graph/graph.rs:199
             at librustc/ty/maps/plumbing.rs:505
             at ./src/librustc_errors/lib.rs:598
  16: rustc::ty::maps::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::cycle_check
             at librustc/ty/maps/plumbing.rs:498
             at librustc/ty/maps/plumbing.rs:121
  17: rustc::ty::maps::<impl rustc::ty::maps::queries::layout_raw<'tcx>>::force
             at librustc/ty/maps/plumbing.rs:497
  18: rustc::ty::maps::<impl rustc::ty::maps::queries::layout_raw<'tcx>>::try_get
             at librustc/ty/maps/plumbing.rs:314
             at librustc/ty/maps/plumbing.rs:539
  19: rustc::ty::maps::TyCtxtAt::layout_raw
             at librustc/ty/maps/plumbing.rs:578
  20: <rustc::ty::layout::LayoutCx<'tcx, rustc::ty::context::TyCtxt<'a, 'tcx, 'tcx>> as rustc::ty::layout::LayoutOf<&'tcx rustc::ty::TyS<'tcx>>>::layout_of
             at librustc/ty/maps/plumbing.rs:571
             at librustc/ty/layout.rs:2052
  21: <&'a rustc::lint::context::LateContext<'a, 'tcx> as rustc::ty::layout::LayoutOf<&'tcx rustc::ty::TyS<'tcx>>>::layout_of
             at librustc/ty/layout.rs:2111
             at librustc/lint/context.rs:634
  22: <rustc_lint::types::VariantSizeDifferences as rustc::lint::LateLintPass<'a, 'tcx>>::check_item
             at librustc_lint/types.rs:781
  23: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::lint::context::LintContext<'tcx>>::with_lint_attrs
             at librustc/lint/context.rs:665
             at librustc/lint/context.rs:625
             at librustc/lint/context.rs:664
             at librustc/lint/context.rs:560
  24: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_mod
             at librustc/lint/context.rs:663
             at librustc/hir/intravisit.rs:183
             at librustc/hir/intravisit.rs:391
             at librustc/lint/context.rs:759
  25: rustc::hir::intravisit::walk_crate
             at librustc/hir/intravisit.rs:377
  26: <rustc::lint::context::LateContext<'a, 'tcx> as rustc::lint::context::LintContext<'tcx>>::with_lint_attrs
             at librustc/lint/context.rs:1040
             at librustc/lint/context.rs:560
  27: rustc::lint::context::check_crate
             at librustc/lint/context.rs:1035
  28: <std::thread::local::LocalKey<T>>::with
             at librustc_driver/driver.rs:1092
             at ./src/librustc/ty/context.rs:1573
             at ./src/libstd/thread/local.rs:377
             at ./src/libstd/thread/local.rs:288
  29: <std::thread::local::LocalKey<T>>::with
             at ./src/librustc/ty/context.rs:1570
             at ./src/librustc/ty/context.rs:1557
             at ./src/libstd/thread/local.rs:377
             at ./src/libstd/thread/local.rs:288
  30: rustc::ty::context::TyCtxt::create_and_enter
             at ./src/librustc/ty/context.rs:1554
             at ./src/librustc/ty/context.rs:1197
  31: rustc_driver::driver::phase_3_run_analysis_passes
             at librustc_driver/driver.rs:1007
  32: rustc_driver::driver::compile_input
             at librustc_driver/driver.rs:214
  33: rustc_driver::run_compiler
             at librustc_driver/lib.rs:506

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.25.0-dev running on x86_64-unknown-linux-gnu


------------------------------------------

thread '[compile-fail] compile-fail/enum-discrim-too-small2.rs' panicked at 'explicit panic', tools/compiletest/src/runtest.rs:2892:9
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:206
   3: std::panicking::default_hook
             at libstd/panicking.rs:216
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:400
   5: std::panicking::begin_panic
   6: compiletest::runtest::ProcRes::fatal
   7: compiletest::runtest::TestCx::fatal_proc_rec
   8: compiletest::runtest::TestCx::check_no_compiler_crash
   9: compiletest::runtest::TestCx::run_cfail_test
  10: compiletest::runtest::run
  11: <F as alloc::boxed::FnBox<A>>::call_box
  12: <F as alloc::boxed::FnBox<A>>::call_box
             at libtest/lib.rs:1456
             at ./src/liballoc/boxed.rs:788
  13: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102


failures:
    [compile-fail] compile-fail/enum-discrim-too-small2.rs

test result: FAILED. 1 passed; 1 failed; 2317 ignored; 0 measured; 0 filtered out

thread 'main' panicked at 'Some tests failed', tools/compiletest/src/main.rs:476:22
stack backtrace:
   0: std::sys::unix::backtrace::tracing::imp::unwind_backtrace
             at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49
   1: std::sys_common::backtrace::print
             at libstd/sys_common/backtrace.rs:71
             at libstd/sys_common/backtrace.rs:59
   2: std::panicking::default_hook::{{closure}}
             at libstd/panicking.rs:206
   3: std::panicking::default_hook
             at libstd/panicking.rs:222
   4: std::panicking::rust_panic_with_hook
             at libstd/panicking.rs:400
   5: std::panicking::begin_panic
   6: compiletest::main
   7: std::rt::lang_start::{{closure}}
   8: std::panicking::try::do_call
             at libstd/rt.rs:59
             at libstd/panicking.rs:305
   9: __rust_maybe_catch_panic
             at libpanic_unwind/lib.rs:102
  10: std::rt::lang_start_internal
             at libstd/panicking.rs:284
             at libstd/panic.rs:361
             at libstd/rt.rs:58
  11: main
  12: __libc_start_main
  13: _start


command did not execute successfully: "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage1/lib" "--run-lib-path" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage1/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/stage1/bin/rustc" "--src-base" "/home/r/src/rust/rustc/src/test/compile-fail" "--build-base" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/test/compile-fail" "--stage-id" "stage1-x86_64-unknown-linux-gnu" "--mode" "compile-fail" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/llvm/build/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Zmiri -Zunstable-options " "--target-rustcflags" "-Crpath -O -Zmiri -Zunstable-options  -Lnative=/home/r/src/rust/rustc/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python" "--lldb-python" "/usr/bin/python" "--gdb" "/usr/bin/gdb" "--llvm-version" "6.0.0\n" "--cc" "" "--cxx" "" "--cflags" "" "--llvm-components" "" "--llvm-cxxflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" ""
expected success, got: exit code: 101
```

I have no idea how my changes could cause that.

**Comment by RalfJung** [comment]
The backtrace points at
```
                bug!("Integer::repr_discr: `#[repr]` hint too small for \
                  discriminant range of enum `{}", ty)
```
which I did not see because it's JSON-ified:
```
{"message":"librustc/ty/layout.rs:553: Integer::repr_discr: `#[repr]` hint too small for discriminant range of enum `Ei64","code":null,"level":"error: internal compiler error","spans":[],"children":[],"rendered":"error: internal compiler error: librustc/ty/layout.rs:553: Integer::repr_discr: `#[repr]` hint too small for discriminant range of enum `Ei64\n\n"}
```
Since this did not happen on Travis, I assume it's a stage-1-only failure. I'm currently building stage 2 to verify.

**Comment by RalfJung** [comment]
> I'm now changing this so the span points at the first ignored bound. Is there a way to "combine" spans? I found no span for all the bounds together, just the generic (which is the wrong place) and individual bounds.

I guess alternatively I could make it warn about every single bound, but that feels excessive?

**Comment by rkruppe** [comment]
You can pass a `MultiSpan`  (or a `Vec<Span>`, there's a From impl) instead of a single span to the `span_*` methods.

File: src/test/compile-fail/issue-39122.rs [hunk 3]
```diff
@@ -8,6 +8,6 @@
 // option. This file may not be copied, modified, or distributed
 // except according to those terms.
 
-type Foo<T: std::ops::Add> = T; //~ WARNING E0122
+type Foo<T: std::ops::Add> = T; //~ WARNING bounds on generic type parameters are ignored
 
-type Bar<T> where T: std::ops::Add = T; //~ WARNING E0122
+type Bar<T> where T: std::ops::Add = T; //~ WARNING where clauses are ignored
```
Comment (RalfJung) [inline src/test/compile-fail/issue-39122.rs]: This test is now a strict subset of the new `ui/param-bounds-ignored`. Should I remove it?

**Comment by RalfJung** [comment]
> You can pass a MultiSpan (or a Vec<Span>, there's a From impl) instead of a single span to the span_* methods.

Wow, fancy. Done :)

**Comment by RalfJung** [comment]
> The warning should be forbid-by-default for cratering though.

I made it Deny-by-default. (There are some existing Deny-by-default lint and no Forbid-by-default lint, so I figured I'd stay consistent with them.)

File: src/test/compile-fail/issue-23046.rs [hunk 4]
```diff
@@ -10,15 +10,15 @@
 
 pub enum Expr<'var, VAR> {
     Let(Box<Expr<'var, VAR>>,
-        Box<for<'v: 'var> Fn(Expr<'v, VAR>) -> Expr<'v, VAR> + 'var>)
```
Comment (petrochenkov) [inline src/test/compile-fail/issue-23046.rs]: Looks like the use of bounds was intentional here (https://github.com/rust-lang/rust/issues/23046).

Comment (petrochenkov) [reply src/test/compile-fail/issue-23046.rs]: Need to check whether they are totally ignored during lowering to `ty` representation or just not enforced properly (and still may cause ICEs or something).

File: src/test/compile-fail/issue-39122.rs (hunk 3 above)
Comment (petrochenkov) [reply src/test/compile-fail/issue-39122.rs]: Yeah, the less code the better!

**Comment by petrochenkov** [comment]
@RalfJung 
>I made it Deny-by-default. (There are some existing Deny-by-default lint and no Forbid-by-default lint, so I figured I'd stay consistent with them.)

For running crater on existing lints forbid-by-default allows to still catch cases where the lint was `allow`ed, but this is a new lint, so there's no difference.

**Comment by petrochenkov** [comment]
Travis found some failing run-pass tests that need to be updated.

Everything else looks good to me, only crater run is remaining.
ping @aidanhs again

**Comment by aidanhs** [comment]
Crater needs to be fed with a try build so it'll need to get through travis, which is unfortunately probably not going to happen with test failures.

(also, please ping @rust-lang/infra for crater runs rather than me directly!)

**Comment by eddyb** [comment]
Regarding https://github.com/rust-lang/rust/pull/48326#issuecomment-366985284, isn't that #47381? cc @pnkfelix

**Comment by aidanhs** [comment]
@bors try

Actually I may be wrong about the try builder running tests - let's find out.

File: src/test/compile-fail/issue-23046.rs (hunk 4 above)
Comment (RalfJung) [reply src/test/compile-fail/issue-23046.rs]: Okay I will bring the bound back. So far I haven't observed anything that would react to these bounds, and AFAIK @eddyb also said they are ignored.

Are you saying I should check the lowering code whether they are ignored? I'm not familiar with that code at all, not even sure where it lives, but I can certainly try.^^

**Comment by aidanhs** [comment]
Crater run started

**Comment by RalfJung** [comment]
Should I preferably rebase or merge to resolve the conflicts?

**Comment by kennytm** [comment]
@RalfJung Rebase is much preferred.

**Comment by RalfJung** [comment]
All right. I'll also squash some commits then to make that less painful.

**Comment by RalfJung** [comment]
Once the crater run is done, where can I see the results?

**Comment by nikomatsakis** [comment]
@RalfJung they get posted on the PR typically. crater, however, can take a good long time!

**Comment by Mark-Simulacrum** [comment]
Crater results: http://cargobomb-reports.s3.amazonaws.com/pr-48326/index.html. 'Blacklisted' crates (spurious failures etc) can be found
[here](https://github.com/rust-lang-nursery/crater/blob/master/blacklist.md).
If you see any spurious failures not on the list, please make a PR against that file.

**Comment by RalfJung** [comment]
I randomly selected some samples from the "regressed" list, and they all look legit -- these are all type aliases that have bounds on them. I have not seen any failures due to bounds in `for` (i.e., higher-ranked lifetimes) yet.

What do we do with this? I kind of expected to make it a warning precisely due to this.

**Comment by petrochenkov** [comment]
I've looked through all the regressions, only `rhusics-ecs-0.1.0` is about `for<'a: 'b> ...`, everything else is either type aliases or (spurious) runtime test failures.

I think we can turn `for<'a: 'b> ...` into an error immediately (`librustc_passes/ast_validation.rs` is a good place for the check). Type aliases can stay a warn-by-default lint.

**Comment by petrochenkov** [comment]
Is there a label S-waiting-on-nikomatsakis?

**Comment by RalfJung** [comment]
> I think we can turn for<'a: 'b> ... into an error immediately (librustc_passes/ast_validation.rs is a good place for the check). Type aliases can stay a warn-by-default lint.

Okay. I will see if I can still make the two checks share some code then.

**Comment by petrochenkov** [comment]
@nikomatsakis 
Is `for<'a: 'b> ...` going to work in the future with Chalk or something?
In this case it may be more reasonable to keep it a lint to avoid unnecessarily breaking something only to fix it later.

**Comment by RalfJung** [comment]
@petrochenkov No longer ignoring these bounds would technically also be a breaking change currently. There's code that compiles right now just because the bounds are ignored. This code's type checking behavior would silently change, were we to stop ignoring the bounds. I'm slightly worried there may be unsafe code that actually thinks it's getting something from these bounds, or unsafe code that is "accidentally" safe just because the bounds are present but ignored -- these are the kinds of bounds unsafe code would use to make sure safe code doesn't go wrong. However, lucky enough, it seems there's not much code out there that uses this bound.

OTOH, if we make this an error now, then we're good -- no code that compiles will be affected by honoring them later.

**Comment by RalfJung** [comment]
> I think we can turn for<'a: 'b> ... into an error immediately (librustc_passes/ast_validation.rs is a good place for the check). 

I think I found a better place: The restriction that `for<T>` is not allowed is realized immediately in the parser; IMHO it makes sense to check both of these restrictions (no type parameters and no bounds) together.

**Comment by RalfJung** [comment]
All right, I made bounds on lifetimes in `for` a hard error, and reduced the lint to bounds on type aliases.

I wasn't entirely sure what to do with `issue-23046.rs` -- it's hard to say now if the original ICE needed the extra bounds. For now I just removed the bounds. Also maybe someone should look over `impl-trait/lifetimes.rs`; I am not sure what exactly that was supposed to test.

**Comment by petrochenkov** [comment]
@RalfJung 
>I think I found a better place

This is not a better place (sorry).
As long as we still keep bounds representable in AST (not everything goes through parser - we still have syntax extensions generating AST directly) and still support empty bound list (`for<'a: /*nothing*/>`) this is not a *syntactic* error.
In general, parser only reports errors for something that can't be preserved in AST and reported later.

**Comment by RalfJung** [comment]
> This is not a better place (sorry).

That's okay. ;)

> In general, parser only reports errors for something that can't be preserved in AST and reported later.

In that case, however, the check for generic type parameters should also be moved elsewhere? The same considerations apply.

**Comment by petrochenkov** [comment]
>In that case, however, the check for generic type parameters should also be moved elsewhere?

If it's an early lint, then it can be moved into AST validation as well (some lints are already reported there).

**Comment by RalfJung** [comment]
I moved them both to AST validation.

Unfortunately, in the parser, there actually is a single place where all the higher-ranked quantifiers are parsed -- whereas in the AST, *some* of the `Vec<GenericParam>` need to be checked (three, if I did not miss any: in bare function types, in poly-trait-refs, and in trait-bound where clauses) while others are not higher-ranked quantifiers and don't need any extra checks.

This makes me wonder if maybe that restriction should be encoded at the type level, i.e. should these places use a `Vec<LateBoundLifetimeParam>` or so instead? That would also make the parser the obvious place to check this; nothing generating the AST directly could even construct an invalid AST.

**Comment by petrochenkov** [comment]
@RalfJung 
>This makes me wonder if maybe that restriction should be encoded at the type level, i.e. should these places use a `Vec<LateBoundLifetimeParam>` or so instead?

AST was closer to `Vec<LateBoundLifetimeParam>` in the past!
But things went into exactly the opposite direction (see https://github.com/rust-lang/rust/pull/45930, including collapsed comments), I'm not especially happy with this either, but the current scheme is indeed more "future-compatible" (whenever that future comes).

**Comment by RalfJung** [comment]
I see. I'd still prefer more precise types here, but I'm sure @eddyb knows this all much better than I do. :)

So, no big refactoring from my side now. IOW, this PR is ready as far as I am concerned.

**Comment by kennytm** [comment]
@bors retry 

Segfault on `dist x86_64-apple-darwin` when building stage2-std for `aarch64-apple-ios` 🤔 cc @oli-obk 

<details><summary>Log</summary>

```
Building stage2 std artifacts (x86_64-apple-darwin -> aarch64-apple-ios)
[01:10:46]    Compiling cc v1.0.4
[01:10:46]    Compiling core v0.0.0 (file:///Users/travis/build/rust-lang/rust/src/libcore)
[01:10:51] error: Could not compile `core`.
[01:10:51] 
[01:10:51] Caused by:
[01:10:51]   process didn't exit successfully: `/Users/travis/build/rust-lang/rust/build/bootstrap/debug/rustc --crate-name core libcore/lib.rs --color always --error-format json --crate-type lib --emit=dep-info,link -C opt-level=2 -C metadata=a37a4ff8d0427b43 -C extra-filename=-a37a4ff8d0427b43 --out-dir /Users/travis/build/rust-lang/rust/build/x86_64-apple-darwin/stage2-std/aarch64-apple-ios/release/deps --target aarch64-apple-ios -L dependency=/Users/travis/build/rust-lang/rust/build/x86_64-apple-darwin/stage2-std/aarch64-apple-ios/release/deps -L dependency=/Users/travis/build/rust-lang/rust/build/x86_64-apple-darwin/stage2-std/release/deps` (signal: 11, SIGSEGV: invalid memory reference)
```

</details>

<details><summary>Stack trace</summary>

```
Thread 1 Crashed:: rustc
0   librustc-d3247632cf1a7ef9.dylib	0x000000010dd91c38 rustc::ich::hcx::_$LT$impl$u20$rustc_data_structures..stable_hasher..HashStable$LT$rustc..ich..hcx..StableHashingContext$LT$$u27$a$GT$$GT$$u20$for$u20$syntax_pos..span_encoding..Span$GT$::hash_stable::hfe4220a2451c76cf + 760
1   librustc-d3247632cf1a7ef9.dylib	0x000000010dd9c9c0 rustc::ich::impls_syntax::_$LT$impl$u20$rustc_data_structures..stable_hasher..HashStable$LT$rustc..ich..hcx..StableHashingContext$LT$$u27$a$GT$$GT$$u20$for$u20$syntax..tokenstream..TokenTree$GT$::hash_stable::h921b1712b6c84e59 + 96
2   librustc-d3247632cf1a7ef9.dylib	0x000000010dd9c78f rustc::ich::impls_syntax::_$LT$impl$u20$rustc_data_structures..stable_hasher..HashStable$LT$rustc..ich..hcx..StableHashingContext$LT$$u27$a$GT$$GT$$u20$for$u20$$u5b$syntax..ast..Attribute$u5d$$GT$::hash_stable::h4a7621726b5465c2 + 2575
3   librustc-d3247632cf1a7ef9.dylib	0x000000010dd63ce9 _$LT$rustc..hir..map..collector..HirItemLike$LT$T$GT$$u20$as$u20$rustc_data_structures..stable_hasher..HashStable$LT$rustc..ich..hcx..StableHashingContext$LT$$u27$hir$GT$$GT$$GT$::hash_stable::hf75827c097bb9aa1 + 201
4   librustc-d3247632cf1a7ef9.dylib	0x000000010dcd4d7c rustc::dep_graph::graph::DepGraph::with_task::hdb3c4cfe4b061c01 + 2764
5   librustc-d3247632cf1a7ef9.dylib	0x000000010dd61332 _$LT$rustc..hir..map..collector..NodeCollector$LT$$u27$a$C$$u20$$u27$hir$GT$$u20$as$u20$rustc..hir..intravisit..Visitor$LT$$u27$hir$GT$$GT$::visit_item::h1a1cdc2886f1c6dd + 338
6   librustc-d3247632cf1a7ef9.dylib	0x000000010dd618f8 _$LT$rustc..hir..map..collector..NodeCollector$LT$$u27$a$C$$u20$$u27$hir$GT$$u20$as$u20$rustc..hir..intravisit..Visitor$LT$$u27$hir$GT$$GT$::visit_item::h1a1cdc2886f1c6dd + 1816
7   librustc-d3247632cf1a7ef9.dylib	0x000000010dd73288 rustc::hir::map::map_crate::hf34ae5166e0d1a2e + 4744
8   librustc_driver-399dd3f73835a2f9.dylib	0x000000010c87fcb3 rustc_driver::driver::compile_input::h3cdd94d9d679771a + 9907
9   librustc_driver-399dd3f73835a2f9.dylib	0x000000010c89cf1f rustc_driver::run_compiler::hf49df0883bdfe282 + 4479
10  librustc_driver-399dd3f73835a2f9.dylib	0x000000010c7c5c96 std::sys_common::backtrace::__rust_begin_short_backtrace::h1b28ee5980b43c33 + 1046
11  libstd-e1de0f43f6dd24ea.dylib 	0x000000010f51c91f __rust_maybe_catch_panic + 31 (lib.rs:110)
12  librustc_driver-399dd3f73835a2f9.dylib	0x000000010c801dd6 _$LT$F$u20$as$u20$alloc..boxed..FnBox$LT$A$GT$$GT$::call_box::h7ca88f6b07c2a955 + 134
13  libstd-e1de0f43f6dd24ea.dylib 	0x000000010f51153c std::sys::unix::thread::Thread::new::thread_start::h655ea7d1eb4459c7 + 140 (thread.rs:25)
14  libsystem_pthread.dylib       	0x00007fff945bd99d _pthread_body + 131
15  libsystem_pthread.dylib       	0x00007fff945bd91a _pthread_start + 168
16  libsystem_pthread.dylib       	0x00007fff945bb351 thread_start + 13
```

</details>

**Comment by Manishearth** [comment]
@bors p=1

segfaulty, not including in rollup, should be tested on its own when there's a slot

**Comment by kennytm** [comment]
@Manishearth The segfault is likely introduced by the miri PR, not this, thus the retry. I agree this (+324, -108) PR shouldn't be rolled up though 😄.

**Comment by RalfJung** [comment]
Sent a PR against the only crate this should break (according to crater): https://github.com/rustgd/rhusics/pull/62

**Pull request merged**